crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,1978 @@
+                if e.im < f64::EPSILON && e.im > -f64::EPSILON {
+                    0.to_string()
+                } else {
+                    String::from(format!("{}i", e.im))
+                }
+            } else if e.im < f64::EPSILON && e.im > -f64::EPSILON {
+                e.re.to_string()
+            } else {
+                e.to_string()
+            },
+        }
+    }
+    pub fn as_real(&self) -> f64 {
+        match self {
+            Value::Real(e) => *e,
+            Value::Int(e) => *e as f64,
+            Value::Complex(e) => e.re,
+        }
+    }
+
+    pub fn abs(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.abs()),
+            Value::Int(e) => Value::Int(e.abs()),
+            Value::Complex(e) => Value::Real((e.re*e.re + e.im*e.im).sqrt()),
+        }
+    }
+    pub fn sin(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.sin()),
+            Value::Int(e) => Value::Real((*e as f64).sin()),
+            Value::Complex(e) => Value::Complex(e.sin()),
+        }
+    }
+    pub fn asin(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.asin()),
+            Value::Int(e) => Value::Real((*e as f64).asin()),
+            Value::Complex(e) => Value::Complex(e.asin()),
+        }
+    }
+    pub fn cos(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.cos()),
+            Value::Int(e) => Value::Real((*e as f64).cos()),
+            Value::Complex(e) => Value::Complex(e.cos()),
+        }
+    }
+    pub fn acos(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.acos()),
+            Value::Int(e) => Value::Real((*e as f64).acos()),
+            Value::Complex(e) => Value::Complex(e.acos()),
+        }
+    }
+    pub fn tan(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.tan()),
+            Value::Int(e) => Value::Real((*e as f64).tan()),
+            Value::Complex(e) => Value::Complex(e.tan()),
+        }
+    }
+    pub fn atan(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.atan()),
+            Value::Int(e) => Value::Real((*e as f64).atan()),
+            Value::Complex(e) => Value::Complex(e.atan()),
+        }
+    }
+    pub fn exp(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.exp()),
+            Value::Int(e) => Value::Real((*e as f64).exp()),
+            Value::Complex(e) => Value::Complex(e.exp()),
+        }
+    }
+    pub fn log(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.ln()),
+            Value::Int(e) => Value::Real((*e as f64).ln()),
+            Value::Complex(e) => Value::Complex(e.ln()),
+        }
+    }
+    pub fn sqrt(&self) -> Value {
+        match self {
+            Value::Real(e) => Value::Real(e.sqrt()),

```

Comment:
what if you give `sqrt` a real negative number? will you get an error or a complex result?